File: src/2024/day2/1/main.go
package main

import (
	_ "embed"
	"fmt"
	"strconv"
	"strings"
)

//go:embed data.txt
var fileContents string

func isLevelValid(level []string) bool {
	// strings to integers
	levelInt := make([]int, len(level))
	for i, l := range level {
		num, err := strconv.Atoi(l)
		if err != nil {
			return false
		}
		levelInt[i] = num
	}

	isIncreasing := true
	isDecreasing := true

	for i := 0; i < len(levelInt)-1; i++ {
		if levelInt[i] >= levelInt[i+1] {
			isIncreasing = false
		}
		if levelInt[i] <= levelInt[i+1] {
			isDecreasing = false
		}
	}

	if !(isIncreasing || isDecreasing) {
		return false
	}

	for i := 0; i < len(levelInt)-1; i++ {
		diff := abs(levelInt[i] - levelInt[i+1])
		if diff < 1 || diff > 3 {
			return false
		}
	}

	return true
}

// abs is a helper function to calculate absolute value
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func process(contents string) int {
	lines := strings.Split(contents, "\n")

	reportCount := 0
	for _, line := range lines {
		levels := strings.Fields(line)
		if isLevelValid(levels) {
			reportCount++
		}
	}

	return reportCount
}

func main() {
	result := process(fileContents)
	fmt.Println(result)
}

File: src/2024/day2/1/main.rs
use std::fs;
use std::path::Path;

fn is_level_valid(level: &[String]) -> bool {
    // strings to integers
    let level_int: Vec<i32> = match level.iter().map(|l| l.parse::<i32>()).collect() {
        Ok(nums) => nums,
        Err(_) => return false,
    };

    let is_increasing = level_int.windows(2).all(|w| w[0] < w[1]);
    let is_decreasing = level_int.windows(2).all(|w| w[0] > w[1]);

    if !(is_increasing || is_decreasing) {
        return false;
    }

    level_int.windows(2).all(|w| {
        let diff = (w[0] - w[1]).abs();
        diff >= 1 && diff <= 3
    })
}

fn read_lines() -> Vec<String> {
    let current_file = Path::new(file!());
    let current_dir = current_file
        .parent()
        .expect("Failed to get parent directory");
    let file_path = current_dir.join("data.txt");

    fs::read_to_string(file_path)
        .expect("Failed to read file")
        .lines()
        .map(|line| line.to_string())
        .filter(|line| !line.is_empty())
        .collect()
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let lines = read_lines();

    let report_count = lines
        .iter()
        .filter(|line| {
            let levels: Vec<String> = line.split_whitespace().map(|s| s.to_string()).collect();
            is_level_valid(&levels)
        })
        .count();

    println!("{}", report_count);
    Ok(())
}